### PR TITLE
advent 2015: add etcd,grpc,http2 article

### PR DESCRIPTION
**Pull request by gyuho**
Reference:
https://docs.google.com/spreadsheets/d/1H0rj7TJo0t8GWyNdGwEc6bGcDj6UgJUN42JtAtbPMrk/edit#gid=0https://docs.google.com/spreadsheets/d/1H0rj7TJo0t8GWyNdGwEc6bGcDj6UgJUN42JtAtbPMrk/edit#gid=0

/cc @xiang90 @jonboulle @philips
Please review and let me know if you have any feedback.
Scheduled posting date is Tuesday, December 29th, 2015.

Thanks,

### CODE REVIEW

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+  ResponseHeader header = 1;
+}
+
+```
+
+And then generate the Go code using `protoc`:
+
+```bash
+#!/usr/bin/env bash
+go get -v -u github.com/gogo/protobuf/proto;
```
Comment (jonboulle) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: You can just go get multiple packages in the one command

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+|AllocedBytesPerOp|3182814152|1795831944|-43.58%|
+
+|RPC|gRPC|gRPC with 100 clients|delta|
+|---|---------|----|:---:|
+|NsPerOp|36716116701|7168591678|-80.48%|
+|AllocsPerOp|25221256|25230286|+0.04%|
+|AllocedBytesPerOp|1795122672|1795831944|+0.04%|
+
+<br>
+As you see, gRPC is much faster and lighter than `jsonrpc`. Not only performant, but also gRPC is easier to reason about concurrency. HTTP/1.x requires multiple TCP connections for concurrent requests, while HTTP/2 can have multiple requests in-flight at the same time over one single TCP connections and still process them asynchronously. Tests above show that gRPC with multiple clients solely speeds up gRPC by 80%, without having to open multiple TCP connections.
```
Comment (jonboulle) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: single TCP connections ->  single TCP connection

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+
+## How does etcd use gRPC?
+
+etcd already uses protocol buffers for all structured data, supporting gRPC as an experimental feature as of today. etcd team is actively working on gRPC integration to replace all etcd RPCs. Here's how etcd is structured to use gRPC:
```
Comment (jonboulle) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: Here's how different parts of the etcd codebase leverage gRPC and Protocol Buffers:

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+     src="https://raw.githubusercontent.com/coreos/etcd/master/logos/etcd-horizontal-color.png"
+     style="float:right; margin-bottom:2em;"/>
+
+etcd is a distributed, consistent key-value store, written in Go. Linux OS uses `/etc` directory to store its system configurations. etcd provides more reliable ways to store such data. It is distributed by replicating data to multiple machines, therefore highly available against single point of failures. Using Raft consensus algorithms, etcd gracefully handles network partitions and machine failures, even leader failures. etcd is being widely used in production: CockroachDB, CoreOS, Kubernetes, etc.
+
+## How does etcd work?
+
+etcd cluster is leader-based. A leader is elected by other members in cluster. Once elected, the leader starts processing client requests and replicating them to its followers. All those server-to-server communications are done by RPC (Remote Procedure Call).
+
+## What matters to etcd?
```
Comment (jonboulle) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: s/etcd/etcd performance/

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+
+Latency matters. Data need to be delivered as fast as possible.  Memory/CPU usage matters. etcd needs to replicate messages with minimum resources. gRPC is one way that can reduce such costs and minimize latency.
```
Comment (jonboulle) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: Maybe you could talk briefly about why these things are important - e.g. lower latency means higher throughput for the state machine, and consequently consistent operations, etc

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+
+<img alt="etcd"
+     src="https://raw.githubusercontent.com/coreos/etcd/master/logos/etcd-horizontal-color.png"
+     style="float:right; margin-bottom:2em;"/>
+
+etcd is a distributed, consistent key-value store, written in Go. Linux OS uses `/etc` directory to store its system configurations. etcd provides more reliable ways to store such data. It is distributed by replicating data to multiple machines, therefore highly available against single point of failures. Using Raft consensus algorithms, etcd gracefully handles network partitions and machine failures, even leader failures. etcd is being widely used in production: CockroachDB, CoreOS, Kubernetes, etc.
+
+## How does etcd work?
+
+etcd cluster is leader-based. A leader is elected by other members in cluster. Once elected, the leader starts processing client requests and replicating them to its followers. All those server-to-server communications are done by RPC (Remote Procedure Call).
```
Comment (jonboulle) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: s/those server-to-server communications/server-to-server communication/

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+title = "etcd: distributed key-value store with grpc/http2"
++++
+
+## What is etcd?
+
+<img alt="etcd"
+     src="https://raw.githubusercontent.com/coreos/etcd/master/logos/etcd-horizontal-color.png"
+     style="float:right; margin-bottom:2em;"/>
+
+etcd is a distributed, consistent key-value store, written in Go. Linux OS uses `/etc` directory to store its system configurations. etcd provides more reliable ways to store such data. It is distributed by replicating data to multiple machines, therefore highly available against single point of failures. Using Raft consensus algorithms, etcd gracefully handles network partitions and machine failures, even leader failures. etcd is being widely used in production: CockroachDB, CoreOS, Kubernetes, etc.
```
Comment (jonboulle) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: Similar to how Linux distributions typically use `/etc` to store local configuration data, etcd can be thought of as a reliable store for _distributed_ configuration data.

**Comment by gyuho** [comment]
@jonboulle Thanks for feedback! I will update the contents based on your comments, and might add more lines about how etcd uses grpc.

**Comment by gyuho** [comment]
Just updated. Still brainstorming what we can add more.

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+title = "etcd: distributed key-value store with grpc/http2"
++++
+
+## What is etcd?
+
+<img alt="etcd"
+     src="https://raw.githubusercontent.com/coreos/etcd/master/logos/etcd-horizontal-color.png"
+     style="float:right; margin-bottom:2em;"/>
+
+etcd is a distributed, consistent key-value store, written in Go. Similar to how Linux distributions typically use `/etc` to store local configuration data, etcd can be thought of as a reliable store for *distributed* configuration data. It is distributed by replicating data to multiple machines, therefore highly available against single point of failures. Using Raft consensus algorithms, etcd gracefully handles network partitions and machine failures, even leader failures. etcd is being widely used in production: CockroachDB, CoreOS, Kubernetes, etc.
```
Comment (philips) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: nit: cockroachdb doesn't use etcd directly but we share a raft library: https://godoc.org/github.com/coreos/etcd/raft

Comment (gyuho) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: I see. Will fix.

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+
+## How does etcd work?
+
+etcd cluster is leader-based. A leader is elected by other members in cluster. Once elected, the leader starts processing client requests and replicating them to its followers. All server-to-server communication is done by RPC (Remote Procedure Call).
```
Comment (philips) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: Perhaps: "etcd clusters are based on a strong leader."

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+
+## What matters to etcd performance?
+
+Latency matters. Data need to be delivered as fast as possible. Lower latency means higher throughput for the state machine, and consequently consistent data replication. Memory/CPU usage matters. etcd needs to replicate messages with minimum resources. gRPC is one way that can reduce such costs and minimize latency.
```
Comment (philips) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: "Data should be delivered..."

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+
+IN PROGRESS...
+
+Will add tests results and code...
+
+```
+
+## Interested?
+
+etcd is still in active development, while widely used in production. If interested, please try it out:
```
Comment (philips) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: I always like to link to the latest release which includes very nice getting started instructions: https://github.com/coreos/etcd/releases/tag/v2.2.2

Comment (gyuho) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: Sure, I will push changes in the next commit as you suggested. Thanks for feedback!

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+
+	api/
+		v3rpc/      // Implements etcd v3 RPC system based on
+					// gRPC and Protocol Buffers.
+
+main.go             // Start 'etcdserver' with grpc framework.
+
+```
+
+## etcd 2.0 vs 2.3+
```
Comment (xiang90) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: probably etcd 2.0 vs etcd 3+ (we will support v3 api officially in etcd3)

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+
+- [etcd v2.0](https://github.com/coreos/etcd/tree/release-2.0) uses Protocol Buffer but does not support gRPC.
+- [etcd v2.3+](https://github.com/coreos/etcd/tree/b1138a42a28e3458ffc738e51f916c2e0c10966c) uses Protocol Buffer with gRPC support.
```
Comment (xiang90) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: uses gPRC with Protocol Buffer to support experimental v3 API.

File: content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md (hunk 12 above)
Comment (gyuho) [inline content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md]: Ok got it.

**Comment by mattfarina** [comment]
Thanks for the post. It's now available at https://blog.gopheracademy.com/advent-2015/etcd-distributed-key-value-store-with-grpc-http2/

**Comment by gyuho** [comment]
Thanks!

**Pull request merged**